### DISCUSSION TITLE
Regarding training object shape and material at once

### DISCUSSION THREAD

**Discussion by yapkap** [mitsuba-renderer/mitsuba3#1859] [Q&A]
Hello,

I'm currently trying to use inverse rendering to train a mesh object, both material/texture as well as vertex positions. In trying to make this happen using mitsuba I've encountered some roadblocks/behaviours I'm in need of help with. I added my current training loop at the end.

Background:
I have a series of 12MP photos which capture an arbitrary object. The area surrounding the object is static and known so I'm only trying to train/learn the object parameters, not any of the surroundings. Because of a weird lighting situation where the scene is essentially equally lit in all directions but also surrounded by an object, I have trained the surrounding geometry with a emitter texture and am also training the object as an area emitter texture.

1. Shape and Textures simultaneously

In going through this discussion https://github.com/mitsuba-renderer/mitsuba3/discussions/600 on the Large Steps optimizer before it was implemeneted in mitsuba natively, there was some mention of the difficulty of solving both material and shape for an object. I could not however find any discussion/issues on the topic since then. I was wondering whether you had any helpful tips/tricks regarding this problem i.e. ideal learning rate ratio or similar. The parameters used in the "Shape optimization" tutorial for example did not work for me.

2. Clipping texture values hinders optimization

I found that when trying to use the clipping operation to valid texture values as used in the "Gradient-based optimization" tutorial, texture values don't change. I only found texture changes when allowing the texture values to exceed the [0,1] interval. Is this expected?

3. Sensor training time

I also found that after the first time a sensor perspective is rendered, subsequent rendering/optimization steps for some but not all sensors can be increased by a static value. This does to my eyes not correspond to the sensor perspective (i.e. a sensor placed directly next to it does not necessarily show the same behaviour) nor does it correspond to the resolution. E.g.:
regular sensor time: 1s --> 4x resolution: 4s
long sensor time: 21s --> 4x resolution: 24s

4. Reference images

Depending on resolution I can currently fit somewhere between 8 and 24 sensor perspectives/their associated images in GPU memory. Is there a way to tell Dr.Jit which images to keep and which to remove from memory? I would want to ideally only hold the current render, its associated reference image and the next 1 or 2 ref images in memory at any given time.

5. Batch sensors

In relation to 4. I've also tried to package all sensors into a combined batch sensor. It did not however show any training progress. I assume this is because of behaviour I encountered when using higher resolution images where I also had to multiply the loss by a value scaled with resolution. I think this is because using MSE the gradients get averaged into oblivion?

```python
# key = 'subject.emitter.radiance.data'
# learning rates: texture: 0.15, vertices: 0.0015
# sensors is a dict not a list
def train_simultaneous(scene, sensors, params, key, args, iterations, ref_path, save_images, out_path, scale, lambda_, transform_dict, samples, untrained_texture, train_texture=True, vertex_train=True):

    learn_rate_vertex = args.vertex_learn_rate
    vertex_start = 2

    used_texture_reset = False
    large_steps_reset = False

    optimizer = mi.ad.Adam(lr=args.learning_rate, mask_updates=True, uniform=True)
    optimizer[key] = params[key]

    params.update(optimizer)

    best_sensors = [1,2,3,4,5,6,7,8,9,10,11,12]

    if vertex_train:
        vertex_positions = dr.unravel(mi.Point3f, params['subject.vertex_positions'])
        vertex_positions[2] = dr.clip(vertex_positions[2], 0.011, 1.0)
        params['subject.vertex_positions'] = dr.ravel(vertex_positions)
        large_steps_t = mi.ad.LargeSteps(params['subject.vertex_positions'], params['subject.faces'], lambda_)
        optimizer['u'] = large_steps_t.to_differential(params['subject.vertex_positions'])
        optimizer.set_learning_rate({key:args.learning_rate, 'u' : learn_rate_vertex})

    print(f"optimizer {optimizer}")

    for it in range(1, iterations+1):

        dr.sync_thread()

        it_start = time.time()

        total_it_loss = mi.Float(0.0)

        num_sens = len(best_sensors)

        # ref_images = [photo_to_drtensor(ref_path + str(best_sensors[i]) + ".jpg", False) for i in range(num_sens)]

        if vertex_train and it % 15 == 0:
            lr_v = optimizer.learning_rate('u')
            new_lr_v = lr_v / 3
            optimizer.set_learning_rate({'u': new_lr_v})
            print(f"optimizer after lowering lr: {optimizer}")

        for sensor_index in range(args.sensor_count):

            sens_start = time.time()

            # Attempt to clip emittance texture values. But even clipping to invalid value interval
            # still inhibits training.
            # optimizer[key] = dr.clip(optimizer[key], -1.0,2.0)

            # Prevents the subject from clipping into a known object at height z = 0.01 
            if vertex_train and it >= vertex_start:
                # params['subject.vertex_positions'] = large_steps_t.from_differential(optimizer['u'])
                vertex_positions = dr.unravel(mi.Point3f, large_steps_t.from_differential(optimizer['u']))
                vertex_positions[2] = dr.clip(vertex_positions[2], 0.011, 1.0)
                params['subject.vertex_positions'] = dr.ravel(vertex_positions)
            
            params.update(optimizer)

            sensor_idx_good = best_sensors[sensor_index]
            
            ref_image = photo_to_drtensor(ref_path + str(sensor_idx_good) + ".jpg", False)

            image = mi.render(scene, params, sensor=sensors[sensor_idx_good], spp=samples, seed=it)

            if save_images:
                mi.util.write_bitmap(out_path+"/image_simul_sen"+str(sensor_idx_good)+"_it"+str(it)+".jpg", image)

            loss = dr.mean(dr.square(image-ref_image))

            total_it_loss += loss

            dr.backward(loss)

            optimizer.step()

            print(f"It {it} - Sensor {sensor_idx_good}: sensor loss = {round(float(loss.numpy()), 2)} --- sensor time = {round(time.time() - sens_start, 2)}")

        print(f"[INFO] Iteration {it}: total loss = {total_it_loss} --- iteration time = {time.time()-it_start}")
```

Any help would be greatly appreciated.
